File: backend/internal/rbac/rbac.go
package rbac

import "strings"

type Permission string

const (
	SettingsManage   Permission = "settings.manage"
	UsersManage      Permission = "users.manage"
	ResourcesScan    Permission = "resources.scan"
	ServersManage    Permission = "servers.manage"
	TerminalConnect  Permission = "terminal.connect"
	TasksManage      Permission = "tasks.manage"
	AuditManage      Permission = "audit.manage"
	AppsManage       Permission = "apps.manage"
	ContainersManage Permission = "containers.manage"
	DatabaseManage   Permission = "database.manage"
	StorageManage    Permission = "storage.manage"
)

var rolePermissions = map[string]map[Permission]struct{}{
	"owner":    allPermissions(),
	"admin":    allPermissions(),
	"operator": permissionSet(ResourcesScan, ServersManage, TerminalConnect, TasksManage, AppsManage, ContainersManage, DatabaseManage, StorageManage),
	"viewer":   permissionSet(),
	"auditor":  permissionSet(),
}

func Allows(role string, permission Permission) bool {
	if permission == "" {
		return true
	}
	permissions, ok := rolePermissions[normalizeRole(role)]
	if !ok {
		return false
	}
	_, ok = permissions[permission]
	return ok
}

func Permissions(role string) []Permission {
	permissions, ok := rolePermissions[normalizeRole(role)]
	if !ok {
		return nil
	}
	out := make([]Permission, 0, len(permissions))
	for permission := range permissions {
		out = append(out, permission)
	}
	return out
}

func ValidRole(role string) bool {
	_, ok := rolePermissions[normalizeRole(role)]
	return ok
}

func NormalizeRole(role string) string {
	return normalizeRole(role)
}

func normalizeRole(role string) string {
	return strings.ToLower(strings.TrimSpace(role))
}

func allPermissions() map[Permission]struct{} {
	return permissionSet(
		SettingsManage,
		UsersManage,
		ResourcesScan,
		ServersManage,
		TerminalConnect,
		TasksManage,
		AuditManage,
		AppsManage,
		ContainersManage,
		DatabaseManage,
		StorageManage,
	)
}

func permissionSet(permissions ...Permission) map[Permission]struct{} {
	out := make(map[Permission]struct{}, len(permissions))
	for _, permission := range permissions {
		out[permission] = struct{}{}
	}
	return out
}

File: web/src/rbac.ts
export const permissions = {
  settingsManage: 'settings.manage',
  usersManage: 'users.manage',
  resourcesScan: 'resources.scan',
  serversManage: 'servers.manage',
  terminalConnect: 'terminal.connect',
  tasksManage: 'tasks.manage',
  auditManage: 'audit.manage',
  appsManage: 'apps.manage',
  containersManage: 'containers.manage',
  databaseManage: 'database.manage',
  storageManage: 'storage.manage',
  credentialsUse: 'credentials.use',
  credentialsManage: 'credentials.manage'
} as const

export type Permission = typeof permissions[keyof typeof permissions]

const rolePermissions: Record<string, Permission[]> = {
  owner: Object.values(permissions),
  admin: Object.values(permissions),
  operator: [
    permissions.resourcesScan,
    permissions.serversManage,
    permissions.terminalConnect,
    permissions.tasksManage,
    permissions.appsManage,
    permissions.containersManage,
    permissions.databaseManage,
    permissions.storageManage,
    permissions.credentialsUse
  ],
  viewer: [],
  auditor: []
}

export function permissionsForRole(role: string) {
  return rolePermissions[role.trim().toLowerCase()]?.slice() ?? []
}

export function normalizePermissions(role: string, granted?: string[]) {
  const values = [
    ...permissionsForRole(role),
    ...(Array.isArray(granted) ? granted : [])
  ]
  return Array.from(new Set(values.filter((value): value is Permission => isPermission(value))))
}

export function isPermission(value: string): value is Permission {
  return (Object.values(permissions) as string[]).includes(value)
}
